### PR TITLE
Actor-to-actor ask for typed

### PR DESCRIPTION
**Pull request by johanandren**
Refs #23770

### CODE REVIEW

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+package akka.actor.typed.scaladsl
+
+import akka.actor.typed.{ ActorRef, TypedAkkaSpec }
+import akka.testkit.typed.TestKit
+import akka.testkit.typed.scaladsl.TestProbe
+
+import scala.util.{ Failure, Success }
+
+class ScalaDslActorContextSpec extends TestKit with TypedAkkaSpec {
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala]: I think it's better to have more specific test classes, like ScalaDslActorContextAskSpe, rather than one that encourage testing everything of ActorContext in same file.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        ctx.ask(pingPong, Ping) {
+          // TODO this doesn't verify that we executed on this actor thread
+          case Success(pong) ⇒ Pong(ctx.self.path.name + "1")
+          // TODO could we have an even shorter variant that just maps success and throws on error
+          case Failure(ex)   ⇒ throw ex
+        }
+
+        import ctx.Ask
+
+        // FIXME needs the dot to work - alternatives?
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala]: I'd say that we shouldn't have `?` for this multi param ask

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala [hunk 3]
```diff
@@ -62,6 +67,16 @@ import scala.concurrent.ExecutionContextExecutor
   override def spawnAdapter[U](f: java.util.function.Function[U, T], name: String): akka.actor.typed.ActorRef[U] =
     internalSpawnAdapter(f.apply, name)
 
+  // FIXME missing implicit error message
+  override def ask[Req, Res](otherActor: ActorRef[Req], createMessage: ActorRef[Res] ⇒ Req)(responseToOwnProtocol: Try[Res] ⇒ T)(implicit responseTimeout: Timeout, classTag: ClassTag[Res]): Unit = {
+    import akka.actor.typed.scaladsl.AskPattern._
+    implicit val scheduler = system.scheduler
+
+    (new Askable(otherActor) ? createMessage).onComplete(res ⇒
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala]: is `Askable` needed? done by `import AskPattern._`?

**Review by patriknw — COMMENTED** [review]
yay, this is the way to go for `ask`

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala (hunk 3 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala]: I got a conflict because of adding our own implicitries here with `?` , but if I remove that everything is fine.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala (hunk 2 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala]: Alright.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ScalaDslActorContextSpec.scala]: I agree, actually already renamed it like that after pushing this

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      }, "ping-pong", Props.empty.withDispatcherFromConfig("ping-pong-dispatcher"))
+
+      val probe = TestProbe[AnyRef]()
+
+      val snitch = Actor.deferred[Pong] { (ctx) ⇒
+
+        // Timeout comes from TypedAkkaSpec
+
+        ctx.ask(pingPong, Ping) {
+          case Success(pong) ⇒ Pong(ctx.self.path.name + "1", Thread.currentThread().getName)
```
Comment (jrudolph) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: I guess it would make sense to also test the behavior when the result type isn't accepted by the asking actor. With this test `Pong` would be accepted anyway.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+          case Failure(ex)   ⇒ throw ex
```
Comment (jrudolph) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: I think we should have the version without `Success` / `Failure` as the default. In the end, the `Success` / `Failure` behavior was particularly thought to be used with ask that returns a `Future`. 

Isn't the only way to fail by replying with `akka.actor.Status.Failure`? Is that even supported in typed so far?

Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: The most common reason and why I think we need it is for when the timeout hits.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala (hunk 4 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: Good idea (it will throw match error in the actor receive).

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala (hunk 5 above)
Comment (jrudolph) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: How about being able to provide a custom timeout message instead? It could have a default message defined that will lead to default ask timeout handling behavior.

Comment (patriknw) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: `Status.Failure` would only be for timeouts, I agree that it's better to define a separate timeout message.

What would default ask timeout handling be? Perhaps only warning log?

Comment (patriknw) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: one edge case that we also might have to handle is if asking an untyped actor and it replies with `Status.Failure`

Comment (jrudolph) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: Isn't the problem then, that you casted the untyped actor to the wrong type? For untyped actors that can reply with `Status.Failure` there should be a way to define the return type manually to `Try[XYZ]`. There could e.g. be an extra overload / method `askOrFail` where the `Try[T]` needs to be handled manually.

Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: Isn't the reason why that (`a.a.Status.Failure`) worked with untyped kind of accidental, I mean, we respond with this type so therefore users could mimic that (I don't think there is any support in the typed ask impl for that right now). 
I think we should encourage users to have explicit protocols if their actors respond on failures.

I was thinking that you pretty much always explicitly want to act on a failed ask to another actor, and that you need context for that. In what scenario don't you care more than that it goes to the log?

**Comment by johanandren** [comment]
PLS BUILD

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala (hunk 5 above)
Comment (patriknw) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: good point, if you don't care you would use `tell`

I think there could still be cases where you want to `ask` from a typed actor to an old existing actor, where you can't change the old actor. I think it's fine to map such failure to the same message as the timeout message.

**Comment by johanandren** [comment]
I think this is done for final review now.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala [hunk 6]
```diff
@@ -62,5 +65,30 @@ class AskSpec extends TestKit("AskSpec") with TypedAkkaSpec with ScalaFutures {
       val answer = noSuchActor ? Foo("bar")
       answer.recover { case _: AskTimeoutException ⇒ "ask" }.futureValue should ===("ask")
     }
+
+    "must transform a replied akka.actor.Status.Failure to a failed future" in {
+      implicit val untypedSystem = akka.actor.ActorSystem("AskSpec-untyped-1")
+      try {
+        // to be honest I'm not too sure this would actually ever happen
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala]: what does this comment mean?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        }
+
+        // FIXME it is nice that it is the actorref you ask but not sure it is justified with the implicit at all WDYT?
+        import ctx.Ask
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: I think that possibility is nice. Would it even be possible to add that implicit automatically when `ctx` is in scope?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorRefAdapter.scala [hunk 8]
```diff
@@ -32,7 +32,8 @@ private[akka] object ActorRefAdapter {
 
   def toUntyped[U](ref: ActorRef[U]): akka.actor.InternalActorRef =
     ref match {
-      case adapter: ActorRefAdapter[_] ⇒ adapter.untyped
+      case adapter: ActorRefAdapter[_]   ⇒ adapter.untyped
+      case system: ActorSystemAdapter[_] ⇒ system.untyped.guardian
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorRefAdapter.scala]: good catch

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala [hunk 9]
```diff
@@ -167,4 +171,34 @@ trait ActorContext[T] {
    */
   def spawnAdapter[U](f: JFunction[U, T]): ActorRef[U]
 
+  /**
+   * Perform a single request-response message interaction with another actor, and transform the messages back to
+   * the protocol of this actor.
+   *
+   * The interaction has a timeout (to avoid resource a resource leak). If the timeout hits without any response it
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala]: resource a resource :)

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala [hunk 10]
```diff
@@ -167,4 +171,34 @@ trait ActorContext[T] {
...
+   * will be passed as an [[akka.pattern.AskTimeoutException]] to the `responseToOwnProtocol` function.
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala]: Perhaps we should use `java.util.concurrent.TimeoutException` instead

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala (hunk 5 above)
Comment (patriknw) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: yes, that is expected since we use the untyped ask underneath, the Status.Failure will fail the Promise, and so on

**Review by patriknw — COMMENTED** [review]
looking good, only some nitpicks
should we try to add a section to the docs together with each PR?

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala [hunk 11]
```diff
@@ -154,4 +157,52 @@ trait ActorContext[T] { this: akka.actor.typed.javadsl.ActorContext[T] ⇒
...
+
+  /**
+   * Implicit decoration of actorrefs to provide syntax sugar for [[ask]]
+   * @see [[Ask#ask]]
+   */
+  implicit final class Ask[Req](otherActor: ActorRef[Req]) {
+    /**
+     * Provides syntax sugar to call [[ask]] directly on an `ActorRef`:
+     * ```
+     * import ctx.Ask
```
Comment (ktoso) [inline akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: Does not seem like there’s a need to have it inside ctx? Does this not make the code clunky since I’m every behavior you will need to import the ask again and again? Could it be external, similar to how in untyped? To enable it in “this file”? Let me know if I missed dissuasion about it though..

Comment (ktoso) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: (Totally +1 for having it though, somewhere)

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala (hunk 6 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala]: It struck me while writing this test: The responding untyped actor would have to cast the incoming `ActorRef[Protocol]` to `ActorRef[akka.actor.Status]` before replying with that `Failure`, when would that ever happen?

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala (hunk 11 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: If it's not inside `ctx` you'd have to take an `ActorContext` as an implicit parameter and cannot before hand know the type of the actor/behavior in the current scope, meaning that you have to mark the entire param list as implicit and possibly be more explicit with type parameters which seemed more clunky to me.

Maybe I'm missing some clever way to do this though.

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala (hunk 10 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala]: I agree, good idea.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala (hunk 7 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorContextAskSpec.scala]: I'll think a bit about that.

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala (hunk 11 above)
Comment (ktoso) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: I see... will try to figure something out

Comment (ktoso) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: Yeah you're right... it's not possible to make it safe from completely the outside since we need the context..

One alternative which may be interesting is:

```
pingPong.ask(ctx, Ping) {
          case Success(pong) ⇒ Pong(ctx.self.path.name + "2", Thread.currentThread().getName)
          case Failure(ex)   ⇒ throw ex
        }```
```

that would not need any imports, but does not passing through the ctx. Personally this looks fine to me, but I wonder that people think. The implicit class wouldbe on ActorRef then,

Comment (jrudolph) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: I thought about that alternative as well but in the end I like the

```
ctx.ask
```

more because you can read it as "Hey my context, ask that other actor", which seems to make sense.

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+ * Single response queries
+ * When an actor need to know that the message was processed before continuing 
+ * To allow an actor to resend if a timely response is not produced
+ * To keep track of outstanding requests and not overwhelm a recipient with messages (simple backpressure)
+ * When some context should be attached to the interaction but the protocol does not support that (request id, what query the response was for)
+ 
+**Problems with ask:**
+
+ * There can only be a single response to one `ask`
+ * When `ask` times out, the receiving actor does not know and may still process it to completion, or even start processing it after the fact
```
Comment (patriknw) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: Another thing could be that it is difficult to get the timeouts right, especially when there is a chain of actors and each is using ask to the next. You want the timeout to be short to have a quick error back to the original requester but you don't want to have too many false positives.

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+ 
+## 1:1 Request-Response with ask between two actors
+ 
+In an interaction where there is a 1:1 mapping between a request and a response we can use @scala[`ActorContext.ask` or the implicitly provided `ActorRef.ask`]@java[`ActorContext.ask`] to interact with another actor.
+
+The interaction has two steps, first we need to construct the outgoing message, to do that we need an @scala[`ActorRef[Response]`]@java[`ActorRef<Response>`] to put as recipient in the outgoing message. The second step is to transform the `Response` or the failure to produce a response, into a message that is part of the protocol of the sending actor.
+
+TODO sample
+
```
Comment (patriknw) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: For scala we should explain the curried syntax in scala (replyTo: ActorRef) as second parameter list in the request message. I think we should first show ordinary `replyTo => Req(replyTo, ...)` and possibly show how the curried syntax can make it look "nicer".

**Review by patriknw — COMMENTED** [review]
Excellent to get this kind of documentation started.

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala (hunk 11 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: Perhaps we should just remove the implicits for now. Can always be added later.

I don't think the `import ctx.Ask` improves readability if you would have to add that line in front of each ask.

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    return Behaviors.immutable((ctx, message) -> {
+      if (message instanceof DisableOutput) {
+        return disabledPrinterBehavior();
+      } else if (message instanceof PrintMe) {
+        System.out.println(((PrintMe) message).message);
+        return Behaviors.same();
+      } else {
+        return Behaviors.ignore();
+      }
+    });
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java]: I had actually forgotten about that one (and was thinking "this is clunky" when I wrote it :D)

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala [hunk 15]
```diff
@@ -62,6 +67,22 @@ import scala.concurrent.ExecutionContextExecutor
   override def spawnAdapter[U](f: java.util.function.Function[U, T], name: String): akka.actor.typed.ActorRef[U] =
     internalSpawnAdapter(f.apply, name)
 
+  // Scala API impl
+  override def ask[Req, Res](otherActor: ActorRef[Req], createMessage: ActorRef[Res] ⇒ Req)(responseToOwnProtocol: Try[Res] ⇒ T)(implicit responseTimeout: Timeout, classTag: ClassTag[Res]): Unit = {
```
Comment (chbatey) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala]: Why do we need the classtag here?

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteContextAskSpec.scala [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        sender ! Pong
+        Behaviors.same
+    }
+  }
+
+  val pingPongKey = ServiceKey[Ping]("ping-pong")
+
+}
+
+class RemoteContextAskSpec extends TestKit(RemoteContextAskSpec.config) with WordSpecLike with Matchers {
```
Comment (chbatey) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteContextAskSpec.scala]: This tests leaks actor system? Can use TypedAkkaSpecWithShutdown

**Comment by johanandren** [comment]
One change we didn't discuss but that made ask feel a bit more consistent on the Scala side: I made also the first function a separate parameter list, this should make inferring the type from the actor-ref always work, and also feels a bit more consistent given that the response mapping is a separate parameter list, wdyt?

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/AskPattern.scala [hunk 17]
```diff
@@ -67,20 +75,22 @@ object AskPattern {
...
+
   private final class PromiseRef[U](target: ActorRef[_], untyped: InternalActorRef, timeout: Timeout) {
 
     // Note: _promiseRef mustn't have a type pattern, since it can be null
     private[this] val (_ref: ActorRef[U], _future: Future[U], _promiseRef) =
       if (untyped.isTerminated)
         (
           adapt.ActorRefAdapter[U](untyped.provider.deadLetters),
-          Future.failed[U](new AskTimeoutException(s"Recipient[$target] had already been terminated.")), null)
+          Future.failed[U](new TimeoutException(s"Recipient[$target] had already been terminated.")), null)
```
Comment (chbatey) [inline akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/AskPattern.scala]: Clean up imports from changing this

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 18]
```diff
@@ -0,0 +1,103 @@
+# Typed Actor Interaction Patterns
+
+Interacting with an Actor in Akka Typed is done through an @scala[`ActorRef[T]`]@java[`ActorRef<T>`] where `T` is the type of messages the actor accepts, also known as "protocol". This ensures that only the right kind of messages can be sent to an actor and also ensures no access to the Actor instance internals is available to anyone else but the Actor itself.
```
Comment (chbatey) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: the "protocol"? Maybe?

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+**Scenarios where request response with tell is useful:**
+
+ * Subscribing to an actor that will send many response messages (of the same protocol) back
+ * When communicating between a parent and its children, where the protocol can be made include the messages for the interaction 
+ * ???
+
+**Problems request-response:**
+
+ * Often the response that the other actor wants to send back is not a part of the protocol the sending actor (see adapted request response or ask)
```
Comment (chbatey) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: maybe: "not part of the sending actor's protocol"

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+**Problems with adapted request-response:**
+
+ * It is hard to detect and that a message request was not delivered or processed (see ask)
+ * Only one adaption can be made per response message type, if a new one is registered the old one is replaced, for example different target actors can't have different adaption if they use the same response types, unless some correlation is encoded in the messages
+ * Unless the protocol already includes a way to provide context, for example a request id that is also sent in the response, it is not possible to tie an interaction to some specific context without introducing a new, separate, actor 
+
+ 
+## 1:1 Request-Response with ask between two actors
+ 
+In an interaction where there is a 1:1 mapping between a request and a response we can use @scala[`ActorContext.ask` or the implicitly provided `ActorRef.ask`]@java[`ActorContext.ask`] to interact with another actor.
```
Comment (chbatey) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: Are we removing the implicitly provided ask?

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+The interaction has two steps, first we need to construct the outgoing message, to do that we need an @scala[`ActorRef[Response]`]@java[`ActorRef<Response>`] to put as recipient in the outgoing message. The second step is to transform the `Response` or the failure to produce a response, into a message that is part of the protocol of the sending actor.
+
+TODO sample
+
+
+**Scenarios where ask is useful:**
+
+ * Single response queries
+ * When an actor need to know that the message was processed before continuing
```
Comment (chbatey) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: needs

File: akka-docs/src/main/paradox/interaction-patterns-typed.md [hunk 22]
```diff
@@ -0,0 +1,103 @@
+# Typed Actor Interaction Patterns
```
Comment (chbatey) [inline akka-docs/src/main/paradox/interaction-patterns-typed.md]: Nice :+1:

**Review by chbatey — COMMENTED** [review]
Looking good :+1:

**Review by patriknw — APPROVED** [review]
LGTM when minor things done

**Comment by johanandren** [comment]
Thank you IntelliJ: 
```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-actor-typed-tests/src/test/java/jdocs/akka/typed/InteractionPatternsTest.java:11:1: package javafx.print does not exist
[error] import javafx.print.Printer;
```

**Review by chbatey — APPROVED** [review]

**Comment by patriknw** [comment]
🍾

**Comment by patriknw** [comment]
Refs https://github.com/akka/akka/issues/24371

**Review by ktoso — COMMENTED** [review]
Late to the LGTMing but LGTM, docs are a great start :)

**Comment by dwijnand** [comment]
Why does the Scala impl make sure to map the result in the actor (and that behaviour is asserted in the spec), but the Java impl doesn't?

Asking because it's equally relevant in my `pipeToSelf` addition.

**Comment by patriknw** [comment]
That must be an oversight. Should be implemented in same way, as in scala impl
